src/tools/non-promotion-reasons.md

```diff
@@ -111,16 +163,546 @@ one or more of the following:
 {{site.alert.end}}
 
 
+### Can't promote `this` {#this}
+
+**The cause:**
+You're trying to promote `this`,
+but type promotion for `this` is not yet supported.
+
+One common `this` promotion scenario is when writing [extension methods][].
+If the [`on` type][] of the extension method is a nullable type,
+you'd want to do a null check to see whether `this` is null:
+
+**Example:**
+
+{:.bad}
+```dart
+extension E on int? {
+  int? get i {
+    return this != null;
+  }
+}
+```
+
+**Message:**
+
+```nocode
+`this` can't be promoted. 
+```
+
+**Solution:**
+
+Create a local variable to hold the value of `this`, then perform the null check.
+
+{:.good}
+```dart
+extension E on int? {
+  int? get i {
+    final self = this;
+    return self != null;
+  }
+}
+```
+
+### Only private fields can be promoted {#private}
+
+**The cause:**
+You're trying to promote a field, but the field is not private.
+
+It’s possible for other libraries in your program
+to override public fields with a getter. Because
+[getters might not return a stable value](#not-field),
+and the compiler can't know what other libraries are doing,
+non-private fields cannot be promoted.
+
+**Example:**
+
+{:.bad}
+```dart
+class C {
+  final int? n; // CONTEXT
+  C(this.n);
+}
+
+test(C c) {
+  if (c.n != null) {
+    print(c.n + 1); // ERROR
+  }
+}
+```
+
+Here, the compiler produces an error message
+explaining that the `+` operator can't be called on `n`
+because it is potentially null.
+
+**Message:**
+
+```nocode
+'n' refers to a public field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Making the field private lets the compiler be sure that no outside libraries
+could possibly override its value, so it's safe to promote.
+
+{:.good}
+```dart
+class C {
+  final int? _n;
+  C(this.n);
+}
+
+test(C c) {
+  if (c.n != null) {
+    print(c.n + 1); // OK
+  }
+}
+```
+
+### Only final fields can be promoted {#final}
+
+**The cause:**
+You're trying to promote a field, but the field is not final.
+
+To the compiler, non-final fields could, in principle,
+be modified any time between the time
+they’re tested and the time they’re used.
+So it's not safe for the compiler to promote a non-final nullable type
+to a non-nullable type.
+
+**Example:**
+
+{:.bad}
+```dart
+class C {
+  int? _mutablePrivateField; // CONTEXT
+  Example(this._mutablePrivateField);
+
+  f() {
+    if (_mutablePrivateField != null) {
+      int i = _mutablePrivateField; // ERROR
+    }
+  }
+}
+```
+
+**Message:**
+
+```nocode
+'mutablePrivateField' refers to a non-final field so it couldn’t be promoted.
+```
+
+**Solution:**
+
+Make the field `final`
+
+{:.good}
+```dart
+class Example {
+  int? final _immutablePrivateField; 
```

Comment:
`final` needs to go before the type, so this line should be `final int? _immutablePrivateField;`.